Build for PS2

AJenbo opened the pull request:
```bash
docker run -it -w /project -v $(pwd):/project ps2dev/ps2dev sh

apk add git make cmake gmp mpc1 mpfr

git config --global --add safe.directory /project

cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/usr/local/ps2dev/ps2sdk/ps2dev.cmake

cmake --build build -j $(nproc)
```

if you need to build the SDK (don't forget to clone and build ps2_drivers) you will also need this:
```bash
apk add musl-dev gcc bash
```

![image](https://user-images.githubusercontent.com/204594/184504100-417fae6a-92a7-4e95-b631-be08bdf7c9c3.png)

### todo
- [x] `ReadOnlyTest()` is creating a folder instead of a file, and fails to delete it, failing next lauch
- [x] Resolution is 640x448 ~~instead of 640x480~~ which is the max PS2 can do 
- https://github.com/ps2dev/ps2sdk/pull/322
- https://github.com/ps2dev/ps2sdk/pull/397
- https://github.com/ps2dev/ps2sdk/pull/398
- https://github.com/ps2dev/ps2sdk/pull/404
- https://github.com/ps2dev/ps2sdk/pull/405
- https://github.com/ps2dev/ps2sdk/pull/406
- https://github.com/ps2dev/ps2sdk/pull/407
- https://github.com/ps2dev/ps2sdk/pull/408
- https://github.com/ps2dev/ps2sdk/pull/410
- [ ] Remove unused player sounds to get below the 2MB audio limit
  - [ ] https://github.com/diasurgical/devilutionX/pull/5820
- [ ] No audio on videos (works via audiolib, maybe just enable that again)
- [ ] Writing saves fails on hardware
- [ ] Wrong sounds are being played in various cases on hardware and there are clicks and pops

### Side note
Initially we tried using SDL1.2, but the version provided by ps2sdk is 1.2.8 and we expect 1.2.10. There is also an issue with there port not correctly implementing Uint64. Solving that wasn't to big of an issue, but the game gets stuck during SDL_JoystickInit, so instead we for now have opted for SDL2 which went much smoother.

AJenbo committed:
Build for PS2

glebm commented:
The CLX stuff is done and PS2 SDK recently received lots of updates, might be a good time to give this another go!

AJenbo commented:
Slight update, the latest build chain fixes the ReadOnlyTest issue as well as remove the need for building the drivers.
Currently the game is crashing for me when loading town.

AJenbo commented:
Everything appears to be working now with streaming audio (currently requires individual files and not using an MPQ)... but the PS2 only allows 16 files to be open at the same time meaning that we can't open all the needed audio files for streaming. So that's something we might have to look in to a solution for.

glebm commented:
It should now allow up to 128 open files

glebm commented on CMake/platforms/ps2.cmake:
```diff
@@ -0,0 +1,41 @@
+# Must use a tiny audio buffer due to RAM constraints.
+set(DEFAULT_AUDIO_BUFFER_SIZE 256)
+
+set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -g0")
```
This should not be needed, stripping takes care of removing debug information and I think `-O3` is the default?

AJenbo committed:
Implement PS2 audio via SPU

This give us 2MB of dedicated audio memory and support for compressed
audio at runtime which is a huge boost to the already limited system
memory

AJenbo committed:
Implement streaming

AJenbo committed:
Fix upscaling on PS2

AJenbo committed:
Disable unused sounds

glebm commented on Source/storm/storm_svid.cpp:
```diff
@@ -6,15 +6,19 @@
 
 #include <SmackerDecoder.h>
 
+#ifndef PS2
 #ifndef NOSOUND
```
Consider doing this instead throughout:

```
#if !defined(NOSOUND) && !defined(PS2)
```

AJenbo replied:
Yeah, the svid portion is still just a the stage where i stubbed everything out to avoid linking with audiolib. I still have to find the best way to bring video sound back up.

Wolf3s replied:
Better:
```
#ifdef __PS2__ //Hack for sound deactivation
#define NOSOUND 1
#endif
```

glebm commented on Source/utils/soundsample.cpp:
```diff
@@ -151,25 +188,76 @@ int SoundSample::SetChunk(ArraySharedPtr<std::uint8_t> fileData, std::size_t dwB
+		size = stream_->size;
+		pitch = stream_->pitch;
+	} else if (!file_path_.empty()) {
+		AssetHandle handle = OpenAsset(file_path_.c_str(), size);
+		if (handle.ok()) {
+			size -= 16;
+			uint32_t buffer[3];
+			if (handle.read(buffer, sizeof(buffer))) {
+				pitch = buffer[2];
+			}
+		}
```
It's a bit surprising to have `GetLength()` open a file.
This should probably be done when the `SoundSample` is constructed.

glebm commented on Source/utils/soundsample.h:
```diff
@@ -96,16 +129,24 @@ class SoundSample final {
+	int channel_ = -1;
+	int pan_ = 0;
+	int volume_ = 100;
+	audsrv_adpcm_t *sampleId_ = nullptr;
+	std::unique_ptr<audsrv_adpcm_t> stream_;
```
Consider a `bool owned_;` instead.

glebm commented on Source/utils/display.cpp:
```diff
@@ -361,7 +361,11 @@ void ReinitializeTexture()
 	auto quality = StrCat(static_cast<int>(*sgOptions.Graphics.scaleQuality));
 	SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, quality.c_str());
 
+#ifndef PS2
 	texture = SDLWrap::CreateTexture(renderer, SDL_PIXELFORMAT_RGB888, SDL_TEXTUREACCESS_STREAMING, gnScreenWidth, gnScreenHeight);
+#else
+	texture = SDLWrap::CreateTexture(renderer, SDL_PIXELFORMAT_ABGR1555, SDL_TEXTUREACCESS_STREAMING, gnScreenWidth, gnScreenHeight);
+#endif
```
Consider doing it more generally by adding a CMake define for `DEVILUTIONX_OUTPUT_TEXTURE_FORMAT`

AJenbo committed:
Only open one file at a time

glebm commented on Source/effects.cpp:
```diff
@@ -737,46 +737,46 @@ TSFX sgSFX[] = {
-/*PS_WARR23*/     { sfx_WARRIOR,               "sfx\\warrior\\warior23.wav",  nullptr },
+/*PS_WARR16B*/    { sfx_STREAM,               "sfx\\warrior\\wario16b.wav",  nullptr },
+/*PS_WARR16C*/    { sfx_STREAM,               "sfx\\warrior\\wario16c.wav",  nullptr },
+/*PS_WARR17*/     { sfx_STREAM,               "sfx\\warrior\\warior17.wav",  nullptr },
```
I don't think this is correct.
This file is "Not a chance" and it is actually used `HeroSpeech::NotAChance`

glebm commented on Source/effects.cpp:
```diff
@@ -737,46 +737,46 @@ TSFX sgSFX[] = {
+/*PS_WARR16B*/    { sfx_STREAM,               "sfx\\warrior\\wario16b.wav",  nullptr },
+/*PS_WARR16C*/    { sfx_STREAM,               "sfx\\warrior\\wario16c.wav",  nullptr },
+/*PS_WARR17*/     { sfx_STREAM,               "sfx\\warrior\\warior17.wav",  nullptr },
+/*PS_WARR18*/     { sfx_STREAM,               "sfx\\warrior\\warior18.wav",  nullptr },
```
I don't think this is correct.
This file is "Not a chance" and it is actually used HeroSpeech::NotAChance

Wolf3s commented:
> ```shell
> docker run -it -w /project -v $(pwd):/project ps2dev/ps2dev sh
> 
> apk add git make cmake gmp mpc1 mpfr
> 
> git config --global --add safe.directory /project
> 
> cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/usr/local/ps2dev/ps2sdk/ps2dev.cmake
> 
> cmake --build build -j $(nproc)
> ```
> 
> if you need to build the SDK (don't forget to clone and build ps2_drivers) you will also need this:
> 
> ```shell
> apk add musl-dev gcc bash
> ```
> 
> ![image](https://user-images.githubusercontent.com/204594/184504100-417fae6a-92a7-4e95-b631-be08bdf7c9c3.png)
> 
> ### todo
> * [x]  `ReadOnlyTest()` is creating a folder instead of a file, and fails to delete it, failing next lauch
> * [x]  Resolution is 640x448 ~instead of 640x480~ which is the max PS2 can do
> * [Implement truncate ps2dev/ps2sdk#322](https://github.com/ps2dev/ps2sdk/pull/322)
> * [Add support for 8bit wave files in adpenc ps2dev/ps2sdk#397](https://github.com/ps2dev/ps2sdk/pull/397)
> * [Implement positional audio in adpcm ps2dev/ps2sdk#398](https://github.com/ps2dev/ps2sdk/pull/398)
> * [audsrv: Fix sterio calculations ps2dev/ps2sdk#404](https://github.com/ps2dev/ps2sdk/pull/404)
> * [Return an error when out of SPU memory ps2dev/ps2sdk#405](https://github.com/ps2dev/ps2sdk/pull/405)
> * [Implement audsrv_free_adpcm ps2dev/ps2sdk#406](https://github.com/ps2dev/ps2sdk/pull/406)
> * [Implement audsrv_is_adpcm_playing ps2dev/ps2sdk#407](https://github.com/ps2dev/ps2sdk/pull/407)
> * [Return channel id when sample is successufully started ps2dev/ps2sdk#408](https://github.com/ps2dev/ps2sdk/pull/408)
> * [Read pitch and channels from header instead of hardcoding them ps2dev/ps2sdk#410](https://github.com/ps2dev/ps2sdk/pull/410)
> * [ ]  Remove unused player sounds to get below the 2MB audio limit
>   
>   * [ ]  [Remove unused Warrior sounds #5820](https://github.com/diasurgical/DevilutionX/pull/5820)
> * [ ]  No audio on videos (works via audiolib, maybe just enable that again)
> * [ ]  Writing saves fails on hardware
> * [ ]  Wrong sounds are being played in various cases on hardware and there are clicks and pops
> 
> ### Side note
> Initially we tried using SDL1.2, but the version provided by ps2sdk is 1.2.8 and we expect 1.2.10. There is also an issue with there port not correctly implementing Uint64. Solving that wasn't to big of an issue, but the game gets stuck during SDL_JoystickInit, so instead we for now have opted for SDL2 which went much smoother.

Hey @AJenbo i've taken your fork and adding new changes.

AJenbo commented:
@Wolf3s Thanks, can you tell me if it solved any of the issues? How is the update working?

Wolf3s commented:
> @Wolf3s Thanks, can you tell me if it solved any of the issues? How is the update working?

Well, i´m having some trouble with sol2 library which is pain in the neck form ps2. I have entered on devilutionX server now.

Wolf3s commented:
> > @Wolf3s Thanks, can you tell me if it solved any of the issues? How is the update working?
> 
> Well, i´m having some trouble with sol2 library which is pain in the neck form ps2. I have entered on devilutionX server now.

Solved it´s lua library compile definition. the only fix was adding ```-DLUA_USE_PS2=1```
